contrib/ibmock/lib.h

```diff
@@ -0,0 +1,129 @@
+/**
+ * Copyright (c) 2024, NVIDIA CORPORATION. All rights reserved.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifndef __LIB_H
+#define __LIB_H
+#include <assert.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <pthread.h>
+#include <sys/uio.h>
+
+
+struct array {
+    void   *data;
+    size_t elem_size;
+    int    count;
+};
+
+pthread_mutex_t global_lock = PTHREAD_MUTEX_INITIALIZER;
+
+static inline void lock(void)
+{
+    pthread_mutex_lock(&global_lock);
+}
+
+static inline void unlock(void)
+{
+    pthread_mutex_unlock(&global_lock);
+}
+
+#define min(_a, _b) ((_a) < (_b) ? (_a) : (_b))
+
+#define container_of(_ptr, _type, _member) \
+    ((_type*)((char*)(_ptr) - (char*)&((_type*)0)->_member))
+
+static inline void array_init(struct array *a, int elem_size)
+{
+    memset(a, 0, sizeof(*a));
+    a->elem_size = elem_size;
+}
+
+static inline void array_cleanup(struct array *a)
+{
+    if (a->data) {
+        free(a->data);
+    }
+    if (a->count) {
+        printf("ibmock: non-empty array (count=%d)\n", a->count);
+    }
+}
+
+static inline void *array_append(struct array *a, void *data, size_t len)
+{
+    void *tmp, *ptr;
+
+    assert(len == a->elem_size);
+    tmp = realloc(a->data, (a->count + 1) * a->elem_size);
```

Comment:
removed malloc(). for efficiency, yes in practice it's only used for around 3 add so I think it's ok.

Reply:
ok, but still we do realloc unconditionally, which looks strange to me.
For example you have an array of 3 elements, then remove all 3 (but capacity remains 3), and then you call append - in this case we don't need to realloc, since there is enough capacity for new element
So maybe we want to check if capacity is enough before growing up the buffer?

Reply:
ah ok, we don't store capacity in array, just the actual size

Reply:
yes, fixed anyways adding total field